### ISSUE TITLE
crc32-mismatch for big requests

### ISSUE THREAD

**Issue by zrzka** [nervous-systems/hildebrand#12]
Just started tracking this issue, not sure where the problem is, but maybe anyone already saw it.

I've got simple lambda function ...

```
(defn- get-participants [org survey]
  (query!
    (lambda/credentials)
    :frank-participant
    {:nyx-organization [:= org]
     :id               [:begins-with (str survey ":")]}
    {:project [:nyx-organization :nyx-user :id :timestamp :employee :participant :answers]}
    {:chan (async/chan 1024)}))

(def ^:export list
  (async-lambda-fn
    (fn [ev ctx]
      (go
        (try
          (if-let [_ (schema/check ParticipantListEventSchema ev)]
            (lambda/bad-request! ctx)
            (let [org (:nyx-organization ev)
                  id (:survey ev)
                  participants (get-participants org id)
                  participants (<! (async/into [] participants))]
              (if (some #(instance? js/Error %) participants)
                (lambda/internal-server-error! ctx)
                (lambda/succeed! ctx {:participants (participants-for-user participants "admin")}))))
          (catch :default _
            (lambda/internal-server-error! ctx)))))))
```

... this function does work perfectly for ~100 participants. But when I try to fetch 500 participants (roughly > 100), I've got crc32-mismatch for almost all requests. When I try this in Python, it does work like a charm even for 1000 participants.

An idea what can be wrong?

**Comment by zrzka** [comment]
Table contains 543 entries matching query criteria. Reproducible scenarios:
- No `:limit` and channel with buffer size `1024`. Result is a list with just `crc32-mismatch` error.
- No `:limit` and channel with buffer size `16384`. Result is a list with just `crc32-mismatch` error.
- `:limit 256` and channel with buffer size `16384`. Result is a list with just `crc32-mismatch` error.
- `:limit 128` and channel with buffer size `16384`. Result contains 512 out of 543 items and 513rd item is `crc32-mismatch` error.
- `:limit 32` and channel with buffer size `16384`. Result contains 512 out of 543 items and 513rd item is `crc32-mismatch` error.
- `:limit 8` and channel with buffer size `16384`. Result contains 543 out of 543 items without `crc32-mismatch` error.

In other words, buffer size is irrelevant (did try even smaller buffers as well). The only relevant part is `:limit` = page size.

One DynamoDB table item is approximately ~2kB.

**Comment by moea** [comment]
On Node, this is what's throwing that error:

``` clj
(defn get-utf8-bytes [s]
  (js/Buffer. s "utf8"))

(defn response-checksum-ok? [{:keys [headers body]}]
  (let [input-crc (some-> headers :x-amz-crc32 js/Number)]
    (or (not input-crc)
        (= (:content-encoding headers) "gzip")
        (= input-crc (.unsigned buffer-crc32 (get-utf8-bytes body))))))
```

So for all un-gzipped (due to a Node awkwardness) responses which have an `x-amz-crc32` header, we're using an external module `buffer-crc32`, passing it a utf8 buffer and doing a straightforward comparison.

Thanks for narrowing it down - the fact that it works with `:limit 8` eliminates a bunch of possible sources of error.  I will try reproducing ASAP.

**Comment by moea** [comment]
(N.B. with approx 2kb item size, 500 is around where the 1mb response limit per query operation would be hit - i.e. Dynamo returns a continuation key and the client transparently issues a request for the remainder of the items.  There may be some issue client-side around how we're handling that case)

**Comment by zrzka** [comment]
Thanks. Yep, that's the reason why I started with `:limit` just to avoid 1MB limit, but it doesn't work even for 16. Just 8 or less, which is ~ <= 16kB. Maybe this is the time when DynamoDB starts gzipping response.

**Comment by zrzka** [comment]
I made some changes to Eulalie to get more info. Issued same request with `:limit 256`.

_Response Headers_
- `:x-amzn-requestid XYZ`
- `:x-amz-crc32 56347694`
- `:content-type application/x-amz-json-1.0`
- `:content-length XYZ`
- `:date XYZ`

All values looks good.

_Response_
- `:status 200`
- `Count 256`
- `:scanned-count 256`
- data looks good in response
- `transform-response-body` values looks good as well

Everything looks good except CRC. Input (= header) is 56347694, but the calculated one is 1532718757.

Either DynamoDB calculates it in a wrong way or Eulalie. I read somewhere, that the CRC32 is calculated from compressed content, not from the decompressed content. Maybe this is the issue? Dunno, still investigating ...

**Comment by zrzka** [comment]
Found it ...
- https://github.com/aws/aws-sdk-java/issues/526#issuecomment-213044240

Quote: _Sorry for the long wait. We asked the service team and they are not planning to change the checksum to uncompressed data. We will soon migrate our underlying Apache Httpclient from 4.3.6 to a newer version 4.5. With that said, we will fix this issue in the new Httpclient by disabling response content automatic decompression for dynamo. I will let you know once that happens._

... so, based on this comment, it looks like that the checksum is calculated from compressed data.

**Comment by moea** [comment]
Thanks, that's a big help - there's a clause in `response-checksum-ok?` which always returns `true` (i.e. checksum is ok) if the response was gzipped server-side - it's not clear why it's not getting hit.  If that's an exhaustive header list you gave, then there isn't a `:content-encoding`.  Will investigate.

**Comment by zrzka** [comment]
Yep, it's exhaustive list and `:content-encoding` is missing. Testing this directly in AWS Console via Test button on my Lambda function. Also Node.js 4.3 is my runtime and there's 1536MB of memory. These are the only differences compared to the default configuration.

**Comment by moea** [comment]
Confirming that on the JVM, I get back a `content-encoding: gzip` header w/ a massive response.  Looking into Node.

**Comment by zrzka** [comment]
Just to be more specific from where I get that the `content-encoding` is missing ... Here's the Eulalie diff ...

```
diff --git a/src/eulalie/core.cljc b/src/eulalie/core.cljc
index 320fede..8aeded2 100644
--- a/src/eulalie/core.cljc
+++ b/src/eulalie/core.cljc
@@ -68,8 +68,8 @@
     retries  :retries
     {:keys [max-retries] :as req} :request :as result}]

-  (if-not (platform/response-checksum-ok? aws-resp)
-    [:error {:type :crc32-mismatch}]
+  (if-let [e (platform/response-checksum-ok? aws-resp)]
+    [:error {:type :crc32-mismatch :message (str (assoc e :headers (:headers aws-resp)))}]
     (if (ok? aws-resp)
       [:ok (transform-response-body aws-resp)]
       (let [error (or (platform/http-response->error (:error aws-resp))
diff --git a/src/eulalie/platform.cljs b/src/eulalie/platform.cljs
index d059f99..b04e068 100644
--- a/src/eulalie/platform.cljs
+++ b/src/eulalie/platform.cljs
@@ -75,9 +75,12 @@

 (defn response-checksum-ok? [{:keys [headers body]}]
   (let [input-crc (some-> headers :x-amz-crc32 js/Number)]
-    (or (not input-crc)
-        (= (:content-encoding headers) "gzip")
-        (= input-crc (.unsigned buffer-crc32 (get-utf8-bytes body))))))
+    (cond
+      (not input-crc) nil
+      (= (:content-encoding headers) "gzip") nil
+      :else (let [calc-crc (.unsigned buffer-crc32 (get-utf8-bytes body))]
+              (if-not (= input-crc calc-crc)
+                {:crc {:input input-crc :calculated calc-crc}})))))

 (defn encode-json [x]
   (.stringify js/JSON (clj->js x)))
```

... and here's the full response of my lambda with these changes ...

```
{
  "errorMessage": "Internal Server Error: [#error {:message \"crc32-mismatch: {:crc {:input 56347694, :calculated 1532718757}, :headers {

:x-amzn-requestid \\\"D3OJRHFHRALQLOU5RF1FC0TGDNVV4KQNSO5AEMVJF66Q9ASUAAJG\\\", 
:x-amz-crc32 \\\"56347694\\\",
:content-type \\\"application/x-amz-json-1.0\\\",
:content-length \\\"300608\\\",
:date \\\"Mon, 11 Jul 2016 16:34:29 GMT\\\"}}\",

:data {:type :crc32-mismatch, :message \"{:crc {:input 56347694, :calculated 1532718757}, :headers {

:x-amzn-requestid \\\"D3OJRHFHRALQLOU5RF1FC0TGDNVV4KQNSO5AEMVJF66Q9ASUAAJG\\\", 
:x-amz-crc32 \\\"56347694\\\",
:content-type \\\"application/x-amz-json-1.0\\\",
:content-length \\\"300608\\\",
:date \\\"Mon, 11 Jul 2016 16:34:29 GMT\\\"}}\"}}]"
}
```

**Comment by moea** [comment]
That makes sense, the presence of the header on the JVM prob has to do w/ the default `accept-encoding` in httpkit vs. Node.

Frustratingly, my first attempts at reproducing this locally couldn't force an error.  With Hildebrand 0.4.3 & Node 5.4.0, in a cljs repl:

``` clj
hildebrand.core> (def ch (async/into [] (hildebrand.channeled/query! (eulalie.creds/env) :eulalie-test-table {:name [:= "joe"]} {:limit 500})))
#'hildebrand.core/ch
1059208 3579007877 3579007877
67822 2361197033 2361197033
hildebrand.core> (go (println (count (async/<! ch))))
#object[cljs.core.async.impl.channels.ManyToManyChannel]
500
```

The columns in the two printed lines are response content length, response `x-amz-crc32` and calculated crc32.  This seems like an adequate test case, right?  There are enough items that two requests are issued in order to retrieve them (note the ~1MB response in the first request).  I'll give the same test a go on Lambda if you confirm you're using Hildebrand 0.4.3.

Note that eulalie isn't sending an accept-encoding header on Node - so while you'd expect AWS not to ever respond w/ gzipped data (and certainly not without setting content-encoding), it may well be that the conditions for when it decides to gzip the response are more complex than say - response size only - and I'm not triggering it in the test above for whatever reason.  One thing you might try:

``` diff
diff --git a/src/eulalie/dynamo.cljc b/src/eulalie/dynamo.cljc
index b3dabd1..808bbdf 100644
--- a/src/eulalie/dynamo.cljc
+++ b/src/eulalie/dynamo.cljc
@@ -21,7 +21,8 @@
    :max-retries 10})

 (defmethod eulalie/prepare-request :eulalie.service/dynamo [req]
-  (util.json/prepare-json-request service-defaults req))
+  (-> (util.json/prepare-json-request service-defaults req)
+      (assoc-in [:headers :accept-encoding] "identity")))

 (defmethod util.json/map-request-keys :eulalie.service/dynamo [{:keys [body]}]
   (json.mapping/transform-request body key-types/request-key-types))
```

Which, according [to the code removed Android AWS SDK change](https://github.com/aws/aws-sdk-android/pull/40/files), may have the API behave differently.

**Comment by zrzka** [comment]
On the phone now, will confirm versions & rest later.

AFAIK it's not reproducible locally if it means local DynamoDB as well, because local DynamoDB doesn't support/use compression/gzip. Did read it today in another issue when I was searching for crc32 & DynamoDB error.

**Comment by moea** [comment]
(I just meant off of Lambda, was targeting AWS Dynamo)

**Comment by zrzka** [comment]
It seems that the `identity` is ignored. Still CRC32 even when I add your change. These headers were send:

```
{
  :content-type "application/x-amz-json-1.0",
  :x-amz-target "DynamoDB_20120810.Query",
  :accept-encoding "identity"
}
```

**Comment by moea** [comment]
OK, I'll give this a go on Node4.3 / Lambda / Hildebrand 0.4.3.

**Comment by zrzka** [comment]
Hmm, interesting ... There's undocummented (?) `gzip` option for node request. I tried to add it:

```
(defn req->node [{{:keys [query host port path]} :endpoint :keys [headers method] :as req}]
  (cond->
      {:host   host
       :path   (cond-> path query (str "?" (url/map->query query)))
       :method method
       :headers headers
       :gzip false}
    port (assoc :port port)))
```

And good & bad news. With `:limit 256` (previously failing in 100% cases) and `:gzip false`, it fails in 50% of cases now and it succeeds in another 50% of cases. Which is interesting, some kind of progress, but not 100% reproducible, which is worse. Or it can be some DynamoDB noise today ...

**Comment by zrzka** [comment]
Okay, taking back my comment ... Probably DynamoDB noise, removed `:gzip false` and it succeeds sometimes today as well.

**Comment by moea** [comment]
Still having trouble reproducing - I can retrieve several MB worth of items with query! on Lambda w/ Node 4.3, against Hildebrand 0.4.3 and 0.4.2.  A couple of questions:
- What region is this happening in?  
- What are the types of the attributes on your item?  With my tests, most of the space I'm taking up is w/ long strings.

**Comment by zrzka** [comment]
Region is `eu-west-1`. Sample redacted item ...

```
{
  "answers": [],
  "employee": {
    "calibration": "Splňuje",
    "degree": "Bc.",
    "email": "redacted",
    "first-name": "redacted",
    "hired-on": "23.9.2013",
    "job-title": "specialista Operations III",
    "last-name": "redacted",
    "name": "redacted",
    "organization-unit": "tým Back Office pro korporátní klientelu",
    "unit-code": "8321023"
  },
  "id": "768-ed23e4226872396c:768-03042332c9674432f0",
  "nyx-organization": "a9f5b232-d352-4c32-2bf-68d532bdbbf3",
  "nyx-user": "47d97239-23ee-4231-b652-59ca93189f23",
  "participant": {
    "attends": true,
    "completed-at": null,
    "invited-at": "2016-05-18T08:27:55.000Z"
  },
  "timestamp": "2016-05-24T20:40:42.596Z"
}
```

`answers` is a list of another maps, ...

**Comment by zrzka** [comment]
Here's bigger one with `answers` ...

```
{
  "answers": [
    {
      "duration": 40.133,
      "employees": [
        "768-edf9e4226873asdc:768-asdasdasdadsasda",
        "768-edf9e4226873asdc:768-asdasdasdadsasda",
        "768-edf9e4226873asdc:768-asdasdasdadsasda"
      ],
      "question": "relationships"
    },
    {
      "duration": 51.658,
      "employees": [
        "768-edf9e4226873asdc:768-asdasdasdadsasda",
        "768-edf9e4226873asdc:768-asdasdasdadsasda",
        "768-edf9e4226873asdc:768-asdasdasdadsasda"
      ],
      "question": "expertise"
    },
    {
      "duration": 42.446,
      "employees": [
        "768-edf9e4226873asdc:768-asdasdasdadsasda",
        "768-edf9e4226873asdc:768-asdasdasdadsasda",
        "768-edf9e4226873asdc:768-asdasdasdadsasda"
      ],
      "question": "vision"
    },
    {
      "duration": 45.515,
      "employees": [
        "768-edf9e4226873asdc:768-asdasdasdadsasda",
        "768-edf9e4226873asdc:768-asdasdasdadsasda",
        "768-edf9e4226873asdc:768-asdasdasdadsasda"
      ],
      "question": "trust"
    },
    {
      "duration": 42.211,
      "employees": [
        "768-edf9e4226873asdc:768-asdasdasdadsasda",
        "768-edf9e4226873asdc:768-asdasdasdadsasda",
        "768-edf9e4226873asdc:768-asdasdasdadsasda"
      ],
      "question": "results"
    }
  ],
  "employee": {
    "calibration": "MID70 - Splňuje",
    "degree": null,
    "email": "redacted",
    "first-name": "redacted",
    "hired-on": "2.3.2024",
    "job-title": "specialista Operations I",
    "last-name": "redacted",
    "name": "redacted",
    "organization-unit": "tým Back Office pro korporátní klientelu",
    "unit-code": "123406"
  },
  "id": "768-asdasd:768-asdasdadasdadsasdasda",
  "nyx-organization": "a9asdasdb32-2352-4cfb-b8bf-68d532bdbbf3",
  "nyx-user": "eec0asdad1-5asd01-4238-8asdc-7e2343bd018e",
  "participant": {
    "attends": true,
    "completed-at": "2016-05-18T12:38:23.628Z",
    "invited-at": "2016-05-18T08:27:49.000Z"
  },
  "timestamp": "2016-05-24T20:40:42.596Z"
}
```

**Comment by zrzka** [comment]
Maybe, not sure, have to test it properly, found a solution.

`http.IncomingMessage` implements `stream.Readable`. And here's the quote from `data` event docs:

> The listener callback will be passed the chunk of data as a string if a default encoding has been specified for the stream using the readable.setEncoding() method; otherwise the data will be passed as a Buffer.

And I don't know why, but it came to my mind automatically, that we're getting UTF8 bytes, calculating checksum, ... but there's no place where we set that we would like to receive UTF8 encoded string. So I did add one line (`eulalie/platform.cljs`) ...

```
(defn request! [{body :body {:keys [protocol] :as u} :endpoint :as req} & [{:keys [chan]}]]
  (let [ch       (or chan (async/chan 10))
        node-req (.request
                  (case protocol "http" http "https" https)
                  (-> req (assoc :endpoint (concretize-port u)) req->node clj->js))]
    (.on node-req "response"
         (fn [resp]
           (if (glossop/error? resp)
             (close-with! ch resp)
             (let [headers (aget resp "headers")
                   status  (aget resp "statusCode")]
               (.setEncoding resp "utf8")  ;; <-------------
```

... and it works. 10 out of 10 requests succeeds. Going to remove all my changes and leave just this line there to check if this is really the fix.

**Comment by zrzka** [comment]
Confirmed ...
- `master` code (both `eulalie` & `hildebrand`)
  - 29 requests failed
  - 1 request succeeded

1 success can be because not all items do contain national characters, there're live data, ... (see below).
- `master` + `(.setEncoding resp "utf8")`
  - 30 requests succeeded

... and it's probably because my data do contain our national characters like ěščřžýáíéĚŠČŘŽÝÁÍÉ... and your don't. This can be the reason why your checksum always worked, but mine not. It looks like it's not gzip, but encoding related.

```
diff --git a/src/eulalie/platform.cljs b/src/eulalie/platform.cljs
index d059f99..6b16ea8 100644
--- a/src/eulalie/platform.cljs
+++ b/src/eulalie/platform.cljs
@@ -32,6 +32,7 @@
              (close-with! ch resp)
              (let [headers (aget resp "headers")
                    status  (aget resp "statusCode")]
+               (.setEncoding resp "utf8")
                (async/put! ch {:headers (js->clj headers :keywordize-keys true)
                                :status status})
                (.on resp "data"  (partial async/put! ch))
```

**Comment by moea** [comment]
Thanks for digging into that - I initially suspected it may be encoding related, but when I saw that your requests were succeeding with smaller limits, I threw out that hypothesis.  I can see now how this might behave differently if "data" is fired more than once, i.e. w/ a large response.

I'll update this issue once I have a snapshot release.

**Comment by zrzka** [comment]
Okay, more tests ...
- raised DynamoDB read capacity just to avoid throttling, ...
- `:limit 256` (543 items in table matching query)
- lambda invoked directly on AWS via CLI (aws-cli)
- one second sleep between requests

_Scenario 1_
- `master` branches
- 476 / 600 requests failed - 80% failure ratio (crc32)

_Scenario 2_
- `master` branches + `(.setEncoding resp "utf8")`
- 0% failure rate - 0 / 600 requests failed

**Comment by zrzka** [comment]
>  I initially suspected it may be encoding related, but when I saw that your requests were succeeding with smaller limits, I threw out that hypothesis.

Yep, that's the remaining weird thing. These 600 requests were made in 10 minutes and table items weren't modified in this case, thus, not sure why the failure ratio isn't 100% in scenario 1.

At least, 0% failure rate in scenario 2 is a good sign.

**Comment by moea** [comment]
Let me know if you have luck with 0.4.4-SNAPSHOT

**Comment by zrzka** [comment]
Okay, increased read capacity and testing with 0.4.4-SNAPSHOT(eulalie) and 0.6.8-SNAPSHOT (hildebrand).

**Comment by zrzka** [comment]
```
Attempt 598/600 ... done, failures so far 0
Attempt 599/600 ... done, failures so far 0
Attempt 600/600 ... done, failures so far 0

Failures 0 / 600 requests
```
- `:limit 256`
- `> 1000` items
- ~2kB each item
- national characters like ěščřžýáíé

Looks good!

**Comment by moea** [comment]
Great! I'll make an actual release once I've released the eulalie SNAPSHOT.

**Comment by moea** [comment]
Pushed 0.4.4
